src/passes/gccrs_rustc_successes.rs

```diff
@@ -0,0 +1,104 @@
+use crate::args::Args;
+use crate::copy_rs_files;
+use crate::error::Error;
+use crate::passes::{Pass, TestCase};
+
+use std::fmt::{Display, Formatter, Result as FmtResult};
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::time::Duration;
+
+use wait_timeout::ChildExt;
+
+pub enum GccrsRustcSuccesses {
+    Full,
+    NoStd,
+    NoCore,
+}
+
+impl Display for GccrsRustcSuccesses {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        let s = match self {
+            GccrsRustcSuccesses::Full => "",
+            GccrsRustcSuccesses::NoStd => "no-std",
+            GccrsRustcSuccesses::NoCore => "no-core",
+        };
+
+        write!(f, "{}", s)
+    }
+}
+
+impl Pass for GccrsRustcSuccesses {
+    fn fetch(&self, args: &Args) -> Result<Vec<PathBuf>, Error> {
+        // FIXME: Do we want more tests than just src/test/ui?
+        let rust_path = &args.rust_path;
+        let ui_tests = rust_path.join("src").join("test").join("ui");
+
+        let out_dir = match self {
+            GccrsRustcSuccesses::Full => args.output_dir.clone(),
+            // We need different output files since we're going to modify files for these
+            GccrsRustcSuccesses::NoStd => args.output_dir.join("no-std"),
+            GccrsRustcSuccesses::NoCore => args.output_dir.join("no-core"),
+        };
+
+        copy_rs_files(&ui_tests, &out_dir, rust_path)
+    }
+
+    fn adapt(&self, args: &Args, file: &Path) -> Result<TestCase, Error> {
+        let test_content = fs::read_to_string(file)?;
+
+        // FIXME: We should also see if the file contains a main function maybe?
+        // To know if we can pass `--crate-type=lib`?
+
+        // We're only interested in successes
+        if test_content.contains("ERROR") {
+            return Ok(TestCase::default());
+        }
+
+        let extra_str = match self {
+            GccrsRustcSuccesses::Full => "",
+            GccrsRustcSuccesses::NoStd => "#![no_std]\n",
+            GccrsRustcSuccesses::NoCore => "#![feature(no_core)]\n#![no_core]\n",
+        };
+
+        fs::write(file, format!("{}{}", extra_str, test_content))?;
+
+        if let GccrsRustcSuccesses::NoStd | GccrsRustcSuccesses::NoCore = self {
+            // We should really think about having rustc as a library here :D
```

Comment:
With ``#![rustc_private]`` you could use the ``RunCompiler`` API. Not sure if you want to do that though. I think just executing ``rustc`` with ``Command`` is good enough.